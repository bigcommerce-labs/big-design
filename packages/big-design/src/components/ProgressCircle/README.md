# Progress Circle

## Examples

<!-- STORY -->

<!-- SOURCE -->

---

### Determinant Progress

Determinant Progress is a known amount of time or completeness for a task to complete. For determinate progress indicators, the fill extends along the path of the indicator until the task complete.

### Indeterminant Progress

Indeterminant Progress is an unknown amount of time for a task to complete. For indeterminate progress indicators, the fill expands and contracts as it travels the length of the path until the tasks is complete.

## Properties & Methods

| Prop      | Type                                         | Default    | Default                              |
| :-------- | :------------------------------------------- | :--------- | :----------------------------------- |
| `error`   | `boolean`                                    |            | Toggles error state                  |
| `percent` | `number`                                     |            | Sets the fill length                 |
| `size`    | `'xSmall' \| 'small' \| 'medium' \| 'large'` | `'medium'` | Sets the size of the progress circle |
