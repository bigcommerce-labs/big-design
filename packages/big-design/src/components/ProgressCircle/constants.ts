export const sizes = {
  large: 80,
  medium: 48,
  small: 32,
  xSmall: 16,
};

export const strokeWidths = {
  large: 8,
  medium: 4,
  small: 4,
  xSmall: 2,
};

export const circumferences = {
  large: (sizes.large / 2 - strokeWidths.large) * 2 * Math.PI,
  medium: (sizes.medium / 2 - strokeWidths.medium) * 2 * Math.PI,
  small: (sizes.small / 2 - strokeWidths.small) * 2 * Math.PI,
  xSmall: (sizes.xSmall / 2 - strokeWidths.xSmall) * 2 * Math.PI,
};
