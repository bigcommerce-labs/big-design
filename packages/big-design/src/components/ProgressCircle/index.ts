export { ProgressCircle, ProgressCircleProps } from './ProgressCircle';
