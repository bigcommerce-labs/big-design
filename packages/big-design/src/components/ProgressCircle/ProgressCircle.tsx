import { CheckCircleIcon, ErrorIcon } from '@bigcommerce/big-design-icons';
import React, { useMemo } from 'react';

import { typedMemo } from '../../utils';

import { CIRCLE_DIMENSIONS } from './constants';
import { StyledCircle, StyledCircleFiller, StyledProgressCircle, StyledText } from './styled';

export interface ProgressCircleProps {
  error?: boolean;
  percent?: number;
  size?: 'xSmall' | 'small' | 'medium' | 'large';
}

export const ProgressCircle: React.FC<ProgressCircleProps> = typedMemo(({ error, percent, size = 'medium' }) => {
  const dimensions = CIRCLE_DIMENSIONS[size];

  const renderedError = useMemo(() => {
    return (
      <ErrorIcon
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={percent ? percent : 0}
        aria-valuetext="Error"
        color="danger"
        role="progressbar"
        size={dimensions}
      />
    );
  }, [percent, dimensions]);

  const renderedSuccess = useMemo(() => {
    return (
      <CheckCircleIcon
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={100}
        color="success"
        role="progressbar"
        size={dimensions}
      />
    );
  }, [dimensions]);

  const renderedCircle = useMemo(() => {
    if (typeof percent !== 'number') {
      return (
        <StyledProgressCircle role="progressbar" size={size}>
          <StyledCircle size={size} />
          <StyledCircleFiller size={size} />
        </StyledProgressCircle>
      );
    }

    if (percent === 100) {
      return renderedSuccess;
    }

    return (
      <StyledProgressCircle
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={percent}
        role="progressbar"
        size={size}
      >
        <StyledCircle size={size} />
        <StyledCircleFiller percent={percent} size={size} />
        {(size === 'large' || size === 'medium') && (
          <StyledText size={size}>{percent ? Math.floor(percent) : 0}%</StyledText>
        )}
      </StyledProgressCircle>
    );
  }, [percent, size, renderedSuccess]);

  return error ? renderedError : renderedCircle;
});
