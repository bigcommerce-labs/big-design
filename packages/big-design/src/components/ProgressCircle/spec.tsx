import 'jest-styled-components';
import React from 'react';
import { render } from 'react-testing-library';

import { ProgressCircle } from './ProgressCircle';

test('render determinant progress circle', () => {
  const { container } = render(<ProgressCircle variant="determinant" percent={50} />);

  expect(container).toMatchSnapshot();
});

test('render completed determinant progress circle', () => {
  const { container } = render(<ProgressCircle isComplete={true} variant="indeterminant" />);

  expect(container).toMatchSnapshot();
});

test('render error determinant progress circle', () => {
  const { container } = render(<ProgressCircle error={true} variant="indeterminant" />);

  expect(container).toMatchSnapshot();
});

test('render indeterminant progress bar', () => {
  const { container } = render(<ProgressCircle variant="indeterminant" />);

  expect(container).toMatchSnapshot();
});

test('render large progress circle', () => {
  const { container } = render(<ProgressCircle size="large" />);

  expect(container).toMatchSnapshot();
});

test('render medium progress circle', () => {
  const { container } = render(<ProgressCircle size="medium" />);

  expect(container).toMatchSnapshot();
});

test('render small progress circle', () => {
  const { container } = render(<ProgressCircle size="small" />);

  expect(container).toMatchSnapshot();
});

test('render xSmall progress circle', () => {
  const { container } = render(<ProgressCircle size="xSmall" />);

  expect(container).toMatchSnapshot();
});
