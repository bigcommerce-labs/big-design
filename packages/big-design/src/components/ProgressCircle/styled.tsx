import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled, { css, keyframes } from 'styled-components';

import { CIRCLE_CIRCUMFERENCES, CIRCLE_DIMENSIONS, CIRCLE_STROKE_WIDTHS } from './constants';
import { ProgressCircleProps } from './ProgressCircle';

export const StyledProgressCircle = styled.svg<ProgressCircleProps>`
  ${({ size, theme }) => css`
    height: ${theme.helpers.remCalc(getDimensions(size))};
    width: ${theme.helpers.remCalc(getDimensions(size))};
  `}
`;

export const StyledCircle = styled.circle.attrs(({ size, theme }: any) => ({
  // rem not usable for circle svg cx, cy, and r values in Safari 14
  cx: theme.helpers.emCalc(getDimensions(size) / 2),
  cy: theme.helpers.emCalc(getDimensions(size) / 2),
  r: theme.helpers.emCalc(getDimensions(size) / 2 - getStrokeWidth(size) / 2),
}))<ProgressCircleProps>`
  fill: transparent;
  stroke-width: ${({ size, theme }) => theme.helpers.remCalc(getStrokeWidth(size))};
  stroke: ${({ theme }) => theme.colors.secondary20};
`;

// @ts-expect-error TODO: Remove after upgrading to styled-components v5
export const StyledCircleFiller = styled(StyledCircle)<ProgressCircleProps>`
  stroke-dasharray: ${({ size }) => getStrokeDashArray(size)};
  stroke: ${({ theme }) => theme.colors.primary};
  transform-origin: 50% 50%;

  ${({ percent, size }) =>
    typeof percent === 'number'
      ? css`
          stroke-dashoffset: ${typeof percent === 'number' ? `${fillLength(percent, size)}` : 0};
          transform: rotate(-90deg);
          transition: stroke-dashoffset 0.35s;
        `
      : css`
          animation: ${spin(size)} 1s ease-out infinite;
          stroke-dashoffset: ${fillLength(0, size)};
          transform: rotate(-90deg);
          transition: stroke-dashoffset 0.35s;
        `};
`;

export const StyledText = styled.text.attrs(() => ({
  dominantBaseline: 'central',
  textAnchor: 'middle',
  x: '50%',
  y: '50%',
}))<ProgressCircleProps>`
  font-size: ${({ size, theme }) =>
    size === 'large' ? theme.typography.fontSize.large : theme.typography.fontSize.small};
  font-weight: ${({ size, theme }) =>
    size === 'large' ? theme.typography.fontWeight.semiBold : theme.typography.fontWeight.regular};
`;

const spin = (size: ProgressCircleProps['size']) => keyframes`
  0% {
    stroke-dashoffset: ${fillLength(0, size) * -1};
    transform: rotate(-90deg);
  }
  50% {
    stroke-dashoffset: ${fillLength(37.5, size) * -1};
  }
  100% {
    stroke-dashoffset: ${fillLength(0, size) * -1};
    transform: rotate(270deg);
  }
`;

function getDimensions(size: ProgressCircleProps['size'] = 'medium') {
  return CIRCLE_DIMENSIONS[size];
}

function getStrokeDashArray(size: ProgressCircleProps['size'] = 'medium') {
  return `${CIRCLE_CIRCUMFERENCES[size]} ${CIRCLE_CIRCUMFERENCES[size]}`;
}

function getStrokeWidth(size: ProgressCircleProps['size'] = 'medium') {
  return CIRCLE_STROKE_WIDTHS[size];
}

function fillLength(percent: number, size: ProgressCircleProps['size'] = 'medium') {
  return CIRCLE_CIRCUMFERENCES[size] - (percent / 100) * CIRCLE_CIRCUMFERENCES[size];
}

StyledProgressCircle.defaultProps = { theme: defaultTheme };
StyledCircle.defaultProps = { theme: defaultTheme };
StyledCircleFiller.defaultProps = { theme: defaultTheme };
StyledText.defaultProps = { theme: defaultTheme };
