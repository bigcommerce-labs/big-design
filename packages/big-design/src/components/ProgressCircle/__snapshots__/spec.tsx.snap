// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render completed determinant progress circle 1`] = `
.c0 {
  color: #2AAB3F;
}

<div>
  <svg
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="100"
    class="c0"
    fill="none"
    height="3rem"
    role="progressbar"
    size="48"
    style="width: 3rem; height: 3rem;"
    viewBox="0 0 24 24"
    width="3rem"
  >
    <title />
    <path
      clip-rule="evenodd"
      d="M12.1 22.101c5.524 0 10.001-4.477 10.001-10 0-5.524-4.477-10.001-10-10.001-5.524 0-10 4.477-10 10 0 5.524 4.476 10.001 10 10.001z"
      fill="currentColor"
      fill-rule="evenodd"
    />
    <path
      clip-rule="evenodd"
      d="M10.997 13.434L8.92 11.339l-1.32 1.333 3.396 3.429L17.6 9.434 16.28 8.1l-5.283 5.334z"
      fill="#fff"
      fill-rule="evenodd"
    />
  </svg>
</div>
`;

exports[`render determinant progress circle 1`] = `
.c0 {
  height: 3rem;
  width: 3rem;
}

.c1 {
  fill: transparent;
  stroke-width: 0.25rem;
  stroke: #ECEEF5;
}

.c2 {
  fill: transparent;
  stroke-width: 0.25rem;
  stroke: #ECEEF5;
  stroke-dasharray: 138.23007675795088 138.23007675795088;
  stroke: #3C64F4;
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  stroke-dashoffset: 69.11503837897544;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  -webkit-transition: stroke-dashoffset 0.35s;
  transition: stroke-dashoffset 0.35s;
}

.c3 {
  font-size: 0.875rem;
  font-weight: 400;
}

<div>
  <svg
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    class="c0"
    role="progressbar"
  >
    <circle
      class="c1"
      cx="1.5rem"
      cy="1.5rem"
      r="1.375rem"
    />
    <circle
      class="c2"
      cx="1.5rem"
      cy="1.5rem"
      r="1.375rem"
    />
    <text
      class="c3"
      dominant-baseline="central"
      text-anchor="middle"
      x="50%"
      y="50%"
    >
      50
      %
    </text>
  </svg>
</div>
`;

exports[`render error determinant progress circle 1`] = `
.c0 {
  color: #DB3643;
}

<div>
  <svg
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="0"
    aria-valuetext="Error"
    class="c0"
    fill="none"
    height="3rem"
    role="progressbar"
    size="48"
    style="width: 3rem; height: 3rem;"
    viewBox="0 0 20 20"
    width="3rem"
  >
    <title />
    <circle
      cx="10"
      cy="10"
      fill="currentColor"
      r="8"
    />
    <path
      d="M11 14.5a.5.5 0 01-.5.5h-1a.5.5 0 01-.5-.5v-1a.5.5 0 01.5-.5h1a.5.5 0 01.5.5v1zM11 10.5a.5.5 0 01-.5.5h-1a.5.5 0 01-.5-.5v-5a.5.5 0 01.5-.5h1a.5.5 0 01.5.5v5z"
      fill="#fff"
    />
  </svg>
</div>
`;

exports[`render indeterminant progress bar 1`] = `
.c0 {
  height: 3rem;
  width: 3rem;
}

.c1 {
  fill: transparent;
  stroke-width: 0.25rem;
  stroke: #ECEEF5;
}

.c2 {
  fill: transparent;
  stroke-width: 0.25rem;
  stroke: #ECEEF5;
  stroke-dasharray: 138.23007675795088 138.23007675795088;
  stroke: #3C64F4;
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  -webkit-animation: dpAsNn 1s ease infinite;
  animation: dpAsNn 1s ease infinite;
  stroke-dashoffset: 138.23007675795088;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  -webkit-transition: stroke-dashoffset 0.35s;
  transition: stroke-dashoffset 0.35s;
}

<div>
  <svg
    class="c0"
    role="progressbar"
  >
    <circle
      class="c1"
      cx="1.5rem"
      cy="1.5rem"
      r="1.375rem"
    />
    <circle
      class="c2"
      cx="1.5rem"
      cy="1.5rem"
      r="1.375rem"
    />
  </svg>
</div>
`;

exports[`render large progress circle 1`] = `
.c0 {
  height: 5rem;
  width: 5rem;
}

.c1 {
  fill: transparent;
  stroke-width: 0.5rem;
  stroke: #ECEEF5;
}

.c2 {
  fill: transparent;
  stroke-width: 0.5rem;
  stroke: #ECEEF5;
  stroke-dasharray: 226.1946710584651 226.1946710584651;
  stroke: #3C64F4;
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  -webkit-animation: ibtQPx 1s ease infinite;
  animation: ibtQPx 1s ease infinite;
  stroke-dashoffset: 226.1946710584651;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  -webkit-transition: stroke-dashoffset 0.35s;
  transition: stroke-dashoffset 0.35s;
}

<div>
  <svg
    class="c0"
    role="progressbar"
  >
    <circle
      class="c1"
      cx="2.5rem"
      cy="2.5rem"
      r="2.25rem"
    />
    <circle
      class="c2"
      cx="2.5rem"
      cy="2.5rem"
      r="2.25rem"
    />
  </svg>
</div>
`;

exports[`render medium progress circle 1`] = `
.c0 {
  height: 3rem;
  width: 3rem;
}

.c1 {
  fill: transparent;
  stroke-width: 0.25rem;
  stroke: #ECEEF5;
}

.c2 {
  fill: transparent;
  stroke-width: 0.25rem;
  stroke: #ECEEF5;
  stroke-dasharray: 138.23007675795088 138.23007675795088;
  stroke: #3C64F4;
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  -webkit-animation: dpAsNn 1s ease infinite;
  animation: dpAsNn 1s ease infinite;
  stroke-dashoffset: 138.23007675795088;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  -webkit-transition: stroke-dashoffset 0.35s;
  transition: stroke-dashoffset 0.35s;
}

<div>
  <svg
    class="c0"
    role="progressbar"
  >
    <circle
      class="c1"
      cx="1.5rem"
      cy="1.5rem"
      r="1.375rem"
    />
    <circle
      class="c2"
      cx="1.5rem"
      cy="1.5rem"
      r="1.375rem"
    />
  </svg>
</div>
`;

exports[`render small progress circle 1`] = `
.c0 {
  height: 2rem;
  width: 2rem;
}

.c1 {
  fill: transparent;
  stroke-width: 0.25rem;
  stroke: #ECEEF5;
}

.c2 {
  fill: transparent;
  stroke-width: 0.25rem;
  stroke: #ECEEF5;
  stroke-dasharray: 87.96459430051421 87.96459430051421;
  stroke: #3C64F4;
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  -webkit-animation: cYXnLG 1s ease infinite;
  animation: cYXnLG 1s ease infinite;
  stroke-dashoffset: 87.96459430051421;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  -webkit-transition: stroke-dashoffset 0.35s;
  transition: stroke-dashoffset 0.35s;
}

<div>
  <svg
    class="c0"
    role="progressbar"
  >
    <circle
      class="c1"
      cx="1rem"
      cy="1rem"
      r="0.875rem"
    />
    <circle
      class="c2"
      cx="1rem"
      cy="1rem"
      r="0.875rem"
    />
  </svg>
</div>
`;

exports[`render xSmall progress circle 1`] = `
.c0 {
  height: 1rem;
  width: 1rem;
}

.c1 {
  fill: transparent;
  stroke-width: 0.125rem;
  stroke: #ECEEF5;
}

.c2 {
  fill: transparent;
  stroke-width: 0.125rem;
  stroke: #ECEEF5;
  stroke-dasharray: 43.982297150257104 43.982297150257104;
  stroke: #3C64F4;
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  -webkit-animation: cWvsXs 1s ease infinite;
  animation: cWvsXs 1s ease infinite;
  stroke-dashoffset: 43.982297150257104;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  -webkit-transition: stroke-dashoffset 0.35s;
  transition: stroke-dashoffset 0.35s;
}

<div>
  <svg
    class="c0"
    role="progressbar"
  >
    <circle
      class="c1"
      cx="0.5rem"
      cy="0.5rem"
      r="0.4375rem"
    />
    <circle
      class="c2"
      cx="0.5rem"
      cy="0.5rem"
      r="0.4375rem"
    />
  </svg>
</div>
`;
